src/hotspot/share/opto/superword.cpp

```diff
@@ -2486,6 +2486,63 @@ VStatus VLoopBody::construct() {
   return VStatus::make_success();
 }
 
+// Returns true if the given operation can be vectorized with "truncation" where the upper bits in the integer do not
+// contribute to the result. This is true for most arithmetic operations, but false for operations such as
+// leading/trailing zero count.
+static bool can_subword_truncate(Node* in, const Type* type) {
+  if (in->is_Load() || in->is_Store() || in->is_Convert() || in->is_Phi()) {
+    return true;
+  }
+
+  int opc = in->Opcode();
+
+  // For shorts and chars, check an additional set of nodes.
+  if (type == TypeInt::SHORT || type == TypeInt::CHAR) {
+    switch (opc) {
+    case Op_ReverseBytesS:
+    case Op_ReverseBytesUS:
+      return true;
+    }
+  }
+
+  switch (opc) {
+  case Op_AddI:
+  case Op_SubI:
+  case Op_MulI:
+  case Op_AndI:
+  case Op_OrI:
+  case Op_XorI:
+    return true;
+  }
+
+#ifdef ASSERT
+  if (VectorNode::is_shift_opcode(opc)) {
+    return false;
+  }
+
+  switch (opc) {
+  case Op_AbsI:
+  case Op_DivI:
+  case Op_MinI:
+  case Op_MaxI:
+  case Op_CMoveI:
+  case Op_RotateRight:
+  case Op_RotateLeft:
+  case Op_PopCountI:
+  case Op_ReverseBytesI:
+  case Op_ReverseI:
+  case Op_CountLeadingZerosI:
+  case Op_CountTrailingZerosI:
+    break;
```

Comment:
```suggestion
  switch (opc) {
  // Cannot be truncated:
  case Op_AbsI:
  case Op_DivI:
  case Op_MinI:
  case Op_MaxI:
  case Op_CMoveI:
  case Op_RotateRight:
  case Op_RotateLeft:
  case Op_PopCountI:
  case Op_ReverseBytesI:
  case Op_ReverseI:
  case Op_CountLeadingZerosI:
  case Op_CountTrailingZerosI:
    return false;
```
Here you did a break in assert. For consistency, let's make it a return.

I would also understand if you did not want to have any return in the assert code, to make sure product and debug do not have different behavior. Up to you.